### DISCUSSION TITLE
Many to One with optional foreign key

### DISCUSSION THREAD

**Discussion by dimitarOnGithub** [sqlalchemy/sqlalchemy#10964] [Usage Questions] [answered]
Hey guys,

New user of SQLAlchemy and first time posting a question, so please let me know if this isn't the right place.

I'm trying to come up with a design of a possible many to one relationship where the child model may have two different different parents; I'm specifically building the design for a Flask app, so please ignore the db.Model inheritance, for example:

```
class Child(db.Model):

  id: Mapped[int] = mapped_column(INTEGER, primary_key=True, autoincrement=True)
  parent_name: Mapped[str] = mapped_column(ForeignKey("parent.name"))


class Parent(db.Model):

  id: Mapped[int] = mapped_column(INTEGER, primary_key=True, autoincrement=True)
  name: Mapped[str] = mapped_column(VARCHAR(255), unique=True, nullable=False)
  children: Mapped[List["Child"]] = relationship()
```

This works perfectly fine, as expected, but I'm finding myself in a situation where I want to preserve this design while also possibly making the Parent optional, as in, a Child may be created, but it doesn't have to necessarily relate to a Parent that exists in the database, without getting an IntegrityError, ie:

```
parent = Parent(name="John Doe")
db.session.add(parent)
db.session.commit()
child = Child(parent_name=parent.name)
db.session.add(child)
db.session.commit()
orphan_child = Child(parent_name="Unknown")
db.session.add(orphan_child)
db.session.commit()
```

Thanks a lot in advance!

**Comment by gordthompson** [comment, marked as answer]
For "making the Parent optional, as in, a Child may be created, but it doesn't have to necessarily relate to a Parent", that would be

```python
    parent_name: Mapped[Optional[str]] = mapped_column(ForeignKey("parent.name"))
```

For "child model may have two different different parents", the simplest case would be two foreign keys

```python
    mother_name: Mapped[Optional[str]] = mapped_column(ForeignKey("parent.name"))
    father_name: Mapped[Optional[str]] = mapped_column(ForeignKey("parent.name"))
```

**Comment by zzzeek** [comment]
so a FOREIGN KEY constraint is only supposed to refer to a. a primary key column, in this case the `Parent.id` column, or b. a column with a UNIQUE constraint on it, which here would be `Parent.name` with UNIQUE on it, however as `Parent` has an integer primary key, that's what you'd want to target.   as @gordthompson seems to have just mentioned as I'm typing, the column with the foreign key can be made `nullable=True` which as he illustrates in SQLAlchemy can be indicated by making the annotation optional.

**Comment by dimitarOnGithub** [comment]
Hey guys, thanks a lot for the suggestions, that worked during my testing, even if I opted for a different approach.

For anyone who might come across this in the future, using Optional as suggested works, but the values can be either NULL or an already existing parent, otherwise the underlying database server (in my case SQL) will throw an `a foreign key constraint fails` error.

Since I wanted my Child to be able to have values for `parent_name` irrelevant to the contents of the Parent table, I had to define the relationship in my table declaration:

```
class Child(db.Model):

    id: Mapped[int] = mapped_column(INTEGER, primary_key=True, autoincrement=True)
    parent_name: Mapped[Optional[str]] = mapped_column(VARCHAR(255), nullable=False)


class Parent(db.Model):

    id: Mapped[int] = mapped_column(INTEGER, primary_key=True, autoincrement=True)
    name: Mapped[str] = mapped_column(VARCHAR(255), unique=True, nullable=False)
    children: Mapped[List["Child"]] = relationship("Child",
                                                   primaryjoin="foreign(Child.parent_name) == Parent.name")

```

That allowed me to create and insert Child objects without the necessary Parent objects existing prior to that:
```
parent = Parent(name="John Doe")
db.session.add(parent)
db.session.commit()

johns_child = Child(parent_name=parent.name)
db.session.add(johns_child)
db.session.commit()

another_child = Child(parent_name="Secret")
db.session.add(another_child)
db.session.commit()

```

And the database representation of the same:
```
> select * from child;
+----+-------------+
| id | parent_name |
+----+-------------+
|  1 | John Doe    |
|  2 | Secret      |
+----+-------------+
2 rows in set (0.001 sec)
```
